### DISCUSSION TITLE
User defined "Session" files to save and reopen multiple tabs

### DISCUSSION THREAD

**Discussion by irmlab** [sumatrapdfreader/sumatrapdf#3045] [Ideas]
The idea is to allow user-defined "sumatra-session.ssf" files (in text format), with a list of files to reopen on tabs
The User can then save and open previous "session" locally or from a session located in the path hierarchy.

The name does not need to be constant. You can manage "MyWorkingFiles1.ssf", "MyWorkingFiles2.ssf" inside a folder.
The SSF file is a simple text file, so people can create it from a script, then spawn to Sumatra from the working directory.

In the GUI, the File menu is added with 2 new options:
-- "Save session" item so the User can save the opened tabs in a "sumatra-session.ssf" file.
-- "Open session"  to reopen a session.

Finally, when starting Sumatra without a file parameter, the initialisation will try to locate a Sumatra-session somewhere.
It starts to look in the working folder, then goes upward in the folder hierarchy. 

This would work for portable or installed versions.

Hope this create a good discussion. The solution is very simple and easy

**Comment by GitHubRulesOK** [comment]
Copied from previous discussions about launching all current tabs but should be very easy to adapt for session history.

"some where\CurrentTabs.cmd"
```
FIND "				FilePath =" "SEE NOTES\SumatraPDF-settings.txt" > %temp%\pdflist.txt
FOR /F "tokens=3,2,1 delims=/" %%i in ("%date%") do set IsoDate=%%k-%%j-%%i
FOR /F "tokens=2,1 delims=: " %%m in ("%time%") do set IsoTime=-%%m-%%n-
FOR /F "tokens=3* skip=2 delims= " %%G IN (%temp%\pdflist.txt) DO echo start "" "the target application\handler.exe" "%%G %%H" >> "some where\%IsoDate%%IsoTime%SumatraPDFSavedSession.cmd

```
To make this much easier for other uses such as simply save all current tabs to a list or a command list.
in pre-release \ SumatraPDF-settings.txt use a simpler external call, change `"some where\` to a folder of your choice e.g. "c:\my scripts\
```
ExternalViewers [
	[
		CommandLine = C:\Windows\System32\cmd.exe /c "some where\CurrentTabs.cmd"
		Name = another app ALL current &Tabs
	]
]
```

**NOTE** to anyone copying/pasting the above command, beware when copying command lines that quotes may be changed from required simple double " marks to non working “curly quoted”

There may need to be an adjustment in the way `IsoDate=` behaves on every machine and country, so currently I get on this device `2022-22-Sat 10-20-56-SumatraPDFSavedSession.cmd` and we do not want that `Sat&space` in the name so will alter for this device
`FOR /F "tokens=4,3,2 delims=/ " %%i in ("%date%") do set IsoDate=%%k-%%i-%%j`
so The numbers needed change  AND add a space at end of delims AND the end of line order (i & j reversed)
Took a few minutes to test but now the session is saved as `2022-10-22-21-21-SumatraPDFSavedSession.cmd`

When current pre-release is running it keeps track of the current tabs in a TabStates [ area towards the end of the SumatraPDF-settings.txt. This was "not always" the case in previous versions, including 3.2.

So now we can search for the current tabs filenames by looking for those with 4 TAB characters. That is what the “Find” part does, so the big gap infront of “FilePath” is important otherwise it will find other entries not currently open.

Where it says "**SEE NOTES\** then YOU need to replace that with the location of the settings file so it might be "C:\users\ *your name* \appdata\local\sumatrapdf (for a more recent installed pre-release version) or if you are using a portable version it could be the folder with the SumatraPDF.exe . The easiest way to check is in Advanced settings using “Save AS” and note where that file is located.

ALSO note where I have put "the target application\handler.exe" as my file handler for the tab set to be reopened (may be images or ebooks) you will need to replace that with wherever your app.exe is located . you can usually find that by looking at the properties of a  shortcut.
**For sumatraPDF sesssion use `"c:\the path to \sumatrapdf.exe" -reuse-instance`  

There are at least two restrictions to avoid with this method

ALL open tabs filenames will be sent to the receiving app so if you have an epub open in a tab that name will also be sent but acrobat should just say it cant open it ?
I had allowed for filenames with spaces in the filename ( previously the reason for %G…) however on revisiting this on another machine I find there seems to be a bug in the way sumatrapdf and windows handle the end part of command line so it is now %G %H and sumatrapdf/windows will add the current tab again to each line !! I cant seem to find out how to avoid that without changing the method (see alternative below) but again acrobat should ignore the extra entry.
The number of tabs that can be processed may depend on how long their filenames are and possibly could fail on those with odd non latin characters. I had no problems with a set of 10 mixed files with reasonable lengths of file path. Best to keep your filename paths short and simple. (filenames with - or _ were not a problem for me, but avoid any with ! or = etc, they may not always work)
Good luck
![image](https://user-images.githubusercontent.com/12754913/197361797-d6cb2ad8-6e4a-414d-adba-47e3fd8d5f70.png)

**Comment by GitHubRulesOK** [comment]
@irmlab 
Windows can run a backup cmd or bat file to run a list of files to restore, so to build a history set of files we just need to export from settings at any time.
As shown above the simplest way is to read settings and write current list. That is what `CurrentTabs.cmd` does (I call it sessions.cmd in picture above). I can run that external to SumatraPDF it does not need to be run internally it can be via shortcut anywhere including desktop.

To run from inside SumatraPDF then you need the `ExternalViewers` part

**Reply by irmlab** [reply]
____
ExternalViewers [
	[
		CommandLine = C:\Windows\System32\cmd.exe /c "some where\CurrentTabs.cmd"
		Name = another app ALL current &Tabs
	]
]
____

I read the pages from Web site about the recent implementations.
But I  tested nothing, yet,  (too much to read and to understand before to known what to do).

I will try soon. 
I came with an Idea and maybe I got a solution ! ;D)

Really, a lot of thanks, @GitHubRulesOK

**Reply by irmlab** [reply]
@GitHubRulesOK
Is there a forum section, or a Web page where people exchange "macros" or programming hacks like yours for SumatraPdf ?
 Thanks
(And maybe some methodical documentation  :D)

**Comment by irmlab** [comment]
Very interesting! You apparently solved a similar problem of mine.
I will surely study your work : I really need some kind of "sessions", And you show there is no need to reinvent the wheel

However, IMHO the issue of this scripting approach is that it is too too complex for an average user. 
The hack may be useful for someone, but it is not the good solution for a "PDF reader" used by common users..

Surely I will take time to learn your approach. The same logic can be adapted to be used from Apache/MS Office, too.
Or I will hack the GUI to add  functions to the File Menu -- If I only have time for that

**Reply by GitHubRulesOK** [reply]
Windows is new DOS (Disk OperationS)
CMDs are new BATch
Applications are complex
Everyday personal tasks are easiest scripted in BATches
Schools favor Pythons but that is yet a slower slithering way to write scripts, and MS suggest slow "powerShell" (both have their uses)

for some everyday tasks, basic batches can have shortcomings, that then require fast machine code.exe, like display a pdf
